app/models/sample.rb

```diff
@@ -328,6 +337,62 @@ def initiate_s3_cp
     LogUtil.log_err_and_airbrake("Failed to upload S3 sample '#{name}' (#{id}): #{e}")
   end
 
+  def initiate_basespace_file_transfer
+    # Only continue if the sample status is CREATED and not yet UPLOADED.
+    return unless status == STATUS_CREATED
+
+    Rails.logger.info("HERE #{basespace_dataset_id} #{basespace_access_token}")
+
+    files = files_for_basespace_dataset(basespace_dataset_id, basespace_access_token)
+
+    raise "Could not fetch Basespace files for dataset #{basespace_dataset_id} for sample '#{name}' (#{id})" if files.blank?
+
+    files.each do |file|
+      max_tries = 3
+      try = 0
+      time_between_tries = 10
+
+      Rails.logger.info("Starting upload of sample '#{name}' (#{id}) file '#{file[:name]}' from Basespace")
+      while try < max_tries
+        success = download_from_basespace_to_s3(file[:download_path], sample_input_s3_path, file[:name])
+
+        if success
+          break
+        else
+          Rails.logger.error("Try ##{try}: Upload of sample '#{name}' (#{id}) file '#{file[:name]}' from Basespace failed")
+          try += 1
+          if try < max_tries
+            Kernel.sleep(time_between_tries)
```

Comment:
generally I would suggest an exponential backoff to avoid the scenario of many workers overwhelming our servers or theirs

Reply:
I had trouble stubbing out sleep in the rspec test without calling it with `Kernel`, strangely enough. I can do exponential backoff.